prow/plugins/blunderbuss/blunderbuss.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package blunderbuss
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	pluginName           = "blunderbuss"
+	defaultReviewerCount = 2
+)
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	var configSuffix string
+	switch config.Blunderbuss.ReviewerCount {
+	case 0:
+		configSuffix = "0 (defaults to 2) reviewers."
+	case 1:
+		configSuffix = "1 reviewer."
+	default:
+		configSuffix = fmt.Sprintf("%d reviewers.", config.Blunderbuss.ReviewerCount)
+	}
+	// Omit the fields [WhoCanUse, Usage, Examples] because this plugin is not triggered by human actions.
+	return &pluginhelp.PluginHelp{
+			Description: "The blunderbuss plugin automatically requests reviews from reviewers when a new PR is created. The reviewers are selected based on the reviewers specified in the OWNERS files that apply to the files modified by the PR.",
+			Config: map[string]string{
+				"": "Blunderbuss is currently configured to request reviews from " + configSuffix,
+			},
+		},
+		nil
+}
+
+// weightMap is a map of user to a weight for that user.
+type weightMap map[string]int64
+
+type ownersClient interface {
+	Reviewers(path string) sets.String
+	LeafReviewers(path string) sets.String
+}
+
+type githubClient interface {
+	RequestReview(org, repo string, number int, logins []string) error
+	GetPullRequestChanges(org, repo string, number int) ([]github.PullRequestChange, error)
+}
+
+func handlePullRequest(pc plugins.PluginClient, pre github.PullRequestEvent) error {
+	if pre.Action != github.PullRequestActionOpened {
+		return nil
+	}
+
+	oc, err := pc.OwnersClient.LoadRepoOwners(pre.Repo.Owner.Login, pre.Repo.Name)
+	if err != nil {
+		return fmt.Errorf("error loading RepoOwners: %v", err)
+	}
+
+	return handle(pc.GitHubClient, oc, pc.Logger, pc.PluginConfig.Blunderbuss.ReviewerCount, &pre)
+}
+
+func handle(ghc githubClient, oc ownersClient, log *logrus.Entry, reviewerCount int, pre *github.PullRequestEvent) error {
+	if reviewerCount == 0 {
+		reviewerCount = defaultReviewerCount
```

Comment:
Could we do this in the config loading step?

Reply:
Fixed.